src/Policy.cpp

```diff
@@ -0,0 +1,252 @@
+                    // trailing values in this must be nan
+                    equal = false;
+                }
+                else if (idx < other.m_values.size() && idx >= m_values.size() &&
+                         !std::isnan(oval->second)) {
+                    // trailing values in other must be nan
+                    equal = false;
+                }
+                else {
+#ifdef GEOPM_DEBUG
+                    throw Exception("Policy: Unexpected gaps in policy data structures.",
+                                    GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+#endif
+                }
+            }
+        }
+        return equal;
+    }
+
+    bool Policy::operator!=(const Policy &other) const
+    {
+        return !(*this == other);
+    }
+
+    std::string Policy::to_string(const std::string &delimiter) const
+    {
+        std::ostringstream temp;
+        for (size_t ii = 0; ii < m_values.size(); ++ii) {
+            if (ii != 0) {
+                temp << delimiter;
+            }
+            if (std::isnan(m_values.at(m_names.at(ii)))) {
+                temp << "NAN";
+            }
+            else {
+                temp << geopm::string_format_double(m_values.at(m_names.at(ii)));
+            }
+        }
+        return temp.str();
+    }
+
+    std::string Policy::to_json(void) const
+    {
+        std::stringstream output_str;
+        std::string policy_value;
+        output_str << "{";
+        for (size_t idx = 0; idx < m_names.size() && m_values.find(m_names.at(idx)) != m_values.end(); ++idx) {
+            if (idx > 0) {
+                output_str << ", ";
+            }
+            if (std::isnan(m_values.at(m_names.at(idx)))) {
+                policy_value = "\"NAN\"";
+            }
+            else {
+                policy_value = geopm::string_format_double(m_values.at(m_names.at(idx)));
+            }
+            output_str << "\"" << m_names.at(idx) << "\": " << policy_value;
+        }
+        output_str << "}";
+        return output_str.str();
+    }
+
+    std::vector<double> Policy::to_vector(void) const
+    {
+        std::vector<double> result(m_values.size());
+        for (size_t ii = 0; ii < m_values.size(); ++ii) {
+#ifdef GEOPM_DEBUG
+        if (ii >= m_names.size() || m_values.find(m_names.at(ii)) == m_values.end()) {
+            throw Exception("Policy::to_vector(): Unexpected gaps in policy data structures.",
+                            GEOPM_ERROR_LOGIC, __FILE__, __LINE__;
+        }
+#endif
+            result[ii] = m_values.at(m_names.at(ii));
+        }
+        return result;
+    }
+
+    void Policy::pad_nan_to(size_t size)
+    {
+        if (size < m_values.size()) {
+            throw Exception("Policy::pad_to_nan(): size of policy cannot be reduced.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        for (size_t ii = m_values.size(); ii < size; ++ii) {
+            m_values[m_names.at(ii)] = NAN;
```

Comment:
Should this be checking whether the values are already set?

E.g. may cause issues when padding to 4 (p3 would be overwritten with NAN): `['p1', 'p2', 'p3', 'p4'] -> {'p1': 1, 'p3': 3}`

Reply:
Also missing check that size < m_names.size().  This seems to work okay but a unit test that checks values don't get overwritten.